consensus/propagation/types/types.go

```diff
@@ -10,6 +10,129 @@ import (
 	"github.com/tendermint/tendermint/types"
 )
 
+// TxMetaData keeps track of the hash of a transaction and its location within the
+// protobuf encoded block.
+type TxMetaData struct {
+	Hash  []byte `protobuf:"bytes,1,opt,name=hash,proto3" json:"hash,omitempty"`
+	Start uint32
+	End   uint32
+}
+
+// ToProto converts TxMetaData to its protobuf representation.
+func (t *TxMetaData) ToProto() *protoprop.TxMetaData {
+	return &protoprop.TxMetaData{
+		Hash:  t.Hash,
+		Start: t.Start,
+		End:   t.End,
+	}
+}
+
+// TxMetaDataFromProto converts a protobuf TxMetaData to its Go representation.
+func TxMetaDataFromProto(t *protoprop.TxMetaData) *TxMetaData {
+	return &TxMetaData{
+		Hash:  t.Hash,
+		Start: t.Start,
+		End:   t.End,
+	}
+}
+
+// ValidateBasic checks if the TxMetaData is valid. It fails if Start > End or
+// if the hash is invalid.
+func (t *TxMetaData) ValidateBasic() error {
+	if t.Start > t.End {
+		return errors.New("TxMetaData: Start > End")
+	}
+
+	return types.ValidateHash(t.Hash)
+}
+
+// CompactBlock contains commitments and metadata for reusing transactions that
+// have already been distributed.
+type CompactBlock struct {
+	BpHash    []byte         `json:"bp_hash,omitempty"`
+	Blobs     []*TxMetaData  `json:"blobs,omitempty"`
+	Signature []byte         `json:"signature,omitempty"`
+	Proposal  types.Proposal `json:"proposal,omitempty"`
+	LastLen   uint32         // length of the last part
```

Comment:
doesn't this need a json representation too?

Reply:
hmm yeah good point

Reply:
didn't do in this PR, but not critical if we forget since we're not using json encoding for anything important